optuna/importance/_base.py

```diff
@@ -63,43 +73,91 @@ def evaluate(
+def _get_info_distribution(distribution: BaseDistribution) -> Dict[str, Any]:
+    info_distribution = {}
+    if isinstance(distribution, CategoricalDistribution):
+        info_distribution = {
+            "is_categorical": True,
+            "distribution": distribution,
+        }
+    elif isinstance(
+        distribution,
+        (
+            DiscreteUniformDistribution,
+            IntLogUniformDistribution,
+            IntUniformDistribution,
+            LogUniformDistribution,
+            UniformDistribution,
+        ),
+    ):
+        info_distribution = {
+            "is_categorical": False,
+            "low": distribution.low,
+            "high": distribution.high,
+        }
+
+    return info_distribution
+
+
+def _get_distributions(study: "Study", params: Optional[List[str]]) -> Dict[str, BaseDistribution]:
     _check_evaluate_args(study, params)
-
-    if params is None:
-        return intersection_search_space(study, ordered_dict=True)
-
-    # New temporary required to pass mypy. Seems like a bug.
-    params_not_none = params
-    assert params_not_none is not None
-
-    # Compute the search space based on the subset of trials containing all parameters.
-    distributions = None
-    for trial in study.trials:
+    info_distributions = None
+    for trial in reversed(study._storage.get_all_trials(study._study_id, deepcopy=False)):
         if trial.state != TrialState.COMPLETE:
             continue
 
-        trial_distributions = trial.distributions
-        if not all(name in trial_distributions for name in params_not_none):
-            continue
+        if params is not None:
+            trial_distributions = trial.distributions
+            if not all(name in trial_distributions for name in params):
+                continue
 
-        if distributions is None:
-            distributions = dict(
-                filter(
-                    lambda name_and_distribution: name_and_distribution[0] in params_not_none,
-                    trial_distributions.items(),
-                )
-            )
+        if info_distributions is None:
+            info_distributions = {}
+            for param_name, param_distribution in trial.distributions.items():
+                if (params is None) or (params is not None and param_name in params):
+                    info_distributions[param_name] = _get_info_distribution(param_distribution)
             continue
 
-        if any(
-            trial_distributions[name] != distribution
-            for name, distribution in distributions.items()
-        ):
+        delete_list = []
+        for param_name, info_distribution in info_distributions.items():
+            if param_name not in trial.distributions:
+                delete_list.append(param_name)
+            else:
+                trial_info_distribution = _get_info_distribution(trial.distributions[param_name])
```

Comment:
It feels a bit like over-engineering introducing a dictionary structure here. Actually, can we keep actual distribution objects instead? We could then perhaps keep updating the lows and highs of those objects during the iteration and return them as they are.